In [77]:
import os
from IPython.display import HTML

import numpy as np
import pandas as pd
from datetime import datetime

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, ConfusionMatrixDisplay
from sklearn.ensemble import RandomForestClassifier


from lazypredict.Supervised import LazyClassifier

# DataSets

In [78]:
rep = r"data"
dataset_total = "df_total_2021_2023.csv"

In [79]:
df_total = pd.read_csv(os.path.join(rep, dataset_total), sep=';')
df_total.head()

Num_Acc id_vehicule  catv  obs  obsm  choc  manv  motor id_usager  \
0  202100000001     201 764     1    0     2     1     1      5   267 638   
1  202100000001     201 765     7    0     9     3    17      1   267 639   
2  202100000002     201 762     7    2     2     1     1      0   267 636   
3  202100000002     201 763     7    0     2     1     9      0   267 637   
4  202100000003     201 761     7    0     1     3     1      1   267 634   

   place  ...  nbv  vosp  prof  plan  surf  infra  situ  vma  day_of_week  \
0      1  ...    2     0     1     1     1      0     1   80            1   
1      1  ...    2     0     1     1     1      0     1   80            1   
2      1  ...    2     0     1     1     1      0     1   80            5   
3      1  ...    2     0     1     1     1      0     1   80            5   
4      1  ...    2     0     1     1     1      0     1   50            3   

              gravity  
0  Blessé hospitalisé  
1             Indemne  
2        Blessé léger  
3  Blessé hospitalisé  
4             Indemne  

[5 rows x 41 columns]

In [80]:
df_total.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 242793 entries, 0 to 242792
Data columns (total 41 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Num_Acc      242793 non-null  int64  
 1   id_vehicule  242793 non-null  object 
 2   catv         242793 non-null  int64  
 3   obs          242793 non-null  int64  
 4   obsm         242793 non-null  int64  
 5   choc         242793 non-null  int64  
 6   manv         242793 non-null  int64  
 7   motor        242793 non-null  int64  
 8   id_usager    242793 non-null  object 
 9   place        242793 non-null  int64  
 10  catu         242793 non-null  int64  
 11  grav         242793 non-null  int64  
 12  sexe         242793 non-null  int64  
 13  an_nais      242793 non-null  float64
 14  trajet       242793 non-null  int64  
 15  secu1        242793 non-null  int64  
 16  jour         242793 non-null  int64  
 17  mois         242793 non-null  int64  
 18  an           242793 non-

In [81]:
dataset_cyclistes = "df_cyclistes_2021_2023.csv"
df_acci_cycl = pd.read_csv(os.path.join(rep, dataset_cyclistes), sep=';')
df_acci_cycl.head() 

Num_Acc id_vehicule  catv  obs  obsm  choc  manv  motor id_usager  \
0  202100000001     201 764     1    0     2     1     1      5   267 638   
1  202100000001     201 765     7    0     9     3    17      1   267 639   
2  202100000006     201 752     1    0     0     5     1      5   267 625   
3  202100000006     201 753     7    0     2     1     1      1   267 626   
4  202100000071     201 623     1    0     0     0     0      0   267 428   

   place  ...  nbv  vosp  prof  plan  surf  infra  situ  vma  day_of_week  \
0      1  ...    2     0     1     1     1      0     1   80            1   
1      1  ...    2     0     1     1     1      0     1   80            1   
2      1  ...    2     0     1     1     1      0     1   50            1   
3      1  ...    2     0     1     1     1      0     1   50            1   
4      1  ...    0     0     1     1     1      0     1   30            3   

              gravity  
0  Blessé hospitalisé  
1             Indemne  
2        Blessé léger  
3             Indemne  
4  Blessé hospitalisé  

[5 rows x 41 columns]

In [82]:
df_acci_cycl.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21547 entries, 0 to 21546
Data columns (total 41 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Num_Acc      21547 non-null  int64  
 1   id_vehicule  21547 non-null  object 
 2   catv         21547 non-null  int64  
 3   obs          21547 non-null  int64  
 4   obsm         21547 non-null  int64  
 5   choc         21547 non-null  int64  
 6   manv         21547 non-null  int64  
 7   motor        21547 non-null  int64  
 8   id_usager    21547 non-null  object 
 9   place        21547 non-null  int64  
 10  catu         21547 non-null  int64  
 11  grav         21547 non-null  int64  
 12  sexe         21547 non-null  int64  
 13  an_nais      21547 non-null  float64
 14  trajet       21547 non-null  int64  
 15  secu1        21547 non-null  int64  
 16  jour         21547 non-null  int64  
 17  mois         21547 non-null  int64  
 18  an           21547 non-null  int64  
 19  hrmn

In [83]:
df_cyclistes_seuls = df_acci_cycl[df_acci_cycl['catv'].isin([1,80])]
df_cyclistes_seuls

Num_Acc  id_vehicule  catv  obs  obsm  choc  manv  motor  \
0      202100000001      201 764     1    0     2     1     1      5   
2      202100000006      201 752     1    0     0     5     1      5   
4      202100000071      201 623     1    0     0     0     0      0   
5      202100000071      201 623     1    0     0     0     0      0   
7      202100000108      201 558     1   12     0     1    14      5   
...             ...          ...   ...  ...   ...   ...   ...    ...   
21538  202300054685  155 583 574     1    0     0     7    14      5   
21539  202300054714  155 583 523     1    0     0     0    23      5   
21542  202300054723  155 583 506     1    0     2     8    13      5   
21543  202300054760  155 583 441     1    0     2     1     1      5   
21546  202300054797  155 583 386     1    0     2     8    19      5   

         id_usager  place  ...  nbv  vosp  prof  plan  surf  infra  situ  vma  \
0          267 638      1  ...    2     0     1     1     1      0     1   80   
2          267 625      1  ...    2     0     1     1     1      0     1   50   
4          267 428      1  ...    0     0     1     1     1      0     1   30   
5          267 429      2  ...    0     0     1     1     1      0     1   30   
7          267 328      1  ...    2     0     2     2     1      2     1   30   
...            ...    ...  ...  ...   ...   ...   ...   ...    ...   ...  ...   
21538  203 721 036      1  ...    2     0     2     1     2      0     1   50   
21539  203 720 967      1  ...    2     0     1     1     1      0     1   50   
21542  203 720 944      1  ...    2     2     1     2     1      0     5   30   
21543  203 720 849      1  ...    2     0     1     1     1      0     1   50   
21546  203 720 770      1  ...    2     0     1     1     1      0     1   30   

       day_of_week             gravity  
0                1  Blessé hospitalisé  
2                1        Blessé léger  
4                3  Blessé hospitalisé  
5                3             Indemne  
7                5                 Tué  
...            ...                 ...  
21538            3  Blessé hospitalisé  
21539            4        Blessé léger  
21542            4        Blessé léger  
21543            2  Blessé hospitalisé  
21546            1        Blessé léger  

[11858 rows x 41 columns]

In [84]:
df_cyclistes_seuls['id_vehicule'].nunique()

9781

In [85]:
df_cyclistes_seuls['id_usager'].nunique()

10367

**10 367 cyclistes impliqués (sur 9781 vélos/trottinettes) mais 11 858 lignes car un accident peut avoir plusieurs lieux**

In [86]:
df_describe = df_cyclistes_seuls.describe()
HTML(df_describe.to_html())


In [87]:
df_cyclistes_seuls.catv.value_counts()

catv
1     10475
80     1383
Name: count, dtype: int64

# Préparation

**Pour le ML des cyclistes, on peut déjà enlever les colonnes :**  
les données qui ne servent que d'identifiant unique :  
- Num_Acc  
- id_vehicule  
- id_usager

les données trop précises/redondantes :
- com (le département suffit)
- lat et long (idem)
- motor
  
les données qui n'influencent pas les features étudiées  
- an
- grav (target gravity suffit)
- catv (on ne regarde que les cyclistes/trottinettes)
- vosp (car on ne sait pas si l'accident a eu lieu sur la voie réservée ou pas)


les données à catégoriser :
toutes sauf :
an_nais
vma
nbv

In [88]:
cols = list(df_cyclistes_seuls.columns)
cols

['Num_Acc',
 'id_vehicule',
 'catv',
 'obs',
 'obsm',
 'choc',
 'manv',
 'motor',
 'id_usager',
 'place',
 'catu',
 'grav',
 'sexe',
 'an_nais',
 'trajet',
 'secu1',
 'jour',
 'mois',
 'an',
 'hrmn',
 'lum',
 'dep',
 'com',
 'agg',
 'int',
 'atm',
 'col',
 'lat',
 'long',
 'catr',
 'circ',
 'nbv',
 'vosp',
 'prof',
 'plan',
 'surf',
 'infra',
 'situ',
 'vma',
 'day_of_week',
 'gravity']

# Dataset Cyclistes

## Encodage des features

In [89]:
features_cat = [ 'catv', 'obs', 'obsm', 'choc', 'manv',  'catu', 'sexe', 'trajet', 'secu1', 'jour', 'mois', 'hrmn', 'lum', 'dep', 'agg', 'int', 'atm',
             'col', 'catr', 'circ', 'prof', 'plan', 'surf', 'infra', 'situ',  'day_of_week']
features_total = features_cat + ['an_nais', 'nbv', 'vma']

df_encoded = pd.get_dummies(df_cyclistes_seuls[features_total],columns=features_cat)
X = df_encoded
y = df_cyclistes_seuls['gravity']


In [90]:
standard_scaler = StandardScaler()
X_scaled = pd.DataFrame(standard_scaler.fit_transform(X), columns=X.columns)
X_scaled.head()

an_nais   nbv   vma  catv_1  catv_80  obs_0  obs_1  obs_2  obs_3  obs_4  \
0     0.98 -0.14  2.04    0.36    -0.36   0.29  -0.18  -0.04  -0.03  -0.03   
1     1.43 -0.14  0.15    0.36    -0.36   0.29  -0.18  -0.04  -0.03  -0.03   
2     1.33 -1.94 -1.12    0.36    -0.36   0.29  -0.18  -0.04  -0.03  -0.03   
3     1.33 -1.94 -1.12    0.36    -0.36   0.29  -0.18  -0.04  -0.03  -0.03   
4     1.33 -0.14 -1.12    0.36    -0.36  -3.46  -0.18  -0.04  -0.03  -0.03   

   ...  situ_5  situ_6  situ_8  day_of_week_0  day_of_week_1  day_of_week_2  \
0  ...   -0.41   -0.14   -0.11          -0.40           2.19          -0.46   
1  ...   -0.41   -0.14   -0.11          -0.40           2.19          -0.46   
2  ...   -0.41   -0.14   -0.11          -0.40          -0.46          -0.46   
3  ...   -0.41   -0.14   -0.11          -0.40          -0.46          -0.46   
4  ...   -0.41   -0.14   -0.11          -0.40          -0.46          -0.46   

   day_of_week_3  day_of_week_4  day_of_week_5  day_of_week_6  
0          -0.43          -0.43          -0.35          -0.32  
1          -0.43          -0.43          -0.35          -0.32  
2           2.31          -0.43          -0.35          -0.32  
3           2.31          -0.43          -0.35          -0.32  
4          -0.43          -0.43           2.84          -0.32  

[5 rows x 340 columns]

## Target en 4 catégories (indemne, blessé léger, blessé hospitalisé , tué)

### Division des données

In [91]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, random_state=36, train_size = 0.8, stratify=y)

print('Proportion des gravités dans l\'ensemble d\'entraînement : ')
display(y_train.value_counts(normalize=True))
print('Proportion des gravités dans l\'ensemble de test : ')
display(y_test.value_counts(normalize=True))

Proportion des gravités dans l'ensemble d'entraînement : 


gravity
Blessé léger         0.66
Blessé hospitalisé   0.24
Indemne              0.07
Tué                  0.04
Name: proportion, dtype: float64

Proportion des gravités dans l'ensemble de test : 


gravity
Blessé léger         0.66
Blessé hospitalisé   0.24
Indemne              0.07
Tué                  0.04
Name: proportion, dtype: float64

### Modèle LogisticRegression

In [92]:
model_rl = LogisticRegression(random_state=36)

# Entraîner le modèle sur les données d'entraînement.
model_rl.fit(X_train, y_train)

# Faire des prédictions sur les ensembles d'entraînement et de test.
y_train_pred = model_rl.predict(X_train)
y_test_pred = model_rl.predict(X_test)

# Calculer et afficher la précision (accuracy) sur les deux ensembles.
print('Précision sur l\'ensemble d\'entraînement : ', accuracy_score(y_train, y_train_pred))
print('Précision sur l\'ensemble de test : ', accuracy_score(y_test, y_test_pred))

HTML(pd.DataFrame(confusion_matrix(y_test, y_test_pred),
             index = model_rl.classes_ + " REAL",
             columns = model_rl.classes_ + " PREDICTED").to_html())

Précision sur l'ensemble d'entraînement :  0.7441492726122707
Précision sur l'ensemble de test :  0.7192242833052277


In [93]:
print(classification_report(y_test, y_test_pred))

                    precision    recall  f1-score   support

Blessé hospitalisé       0.57      0.44      0.50       562
      Blessé léger       0.77      0.88      0.82      1559
           Indemne       0.65      0.37      0.47       167
               Tué       0.37      0.19      0.25        84

          accuracy                           0.72      2372
         macro avg       0.59      0.47      0.51      2372
      weighted avg       0.70      0.72      0.70      2372



### Modèle RandomForestClassifier

In [94]:
model_rfc = RandomForestClassifier(random_state=36)

# Entraîner le modèle sur les données d'entraînement.
model_rfc.fit(X_train, y_train)

# Faire des prédictions sur les ensembles d'entraînement et de test.
y_train_pred = model_rfc.predict(X_train)
y_test_pred = model_rfc.predict(X_test)

# Calculer et afficher la précision (accuracy) sur les deux ensembles.
print('Précision sur l\'ensemble d\'entraînement : ', accuracy_score(y_train, y_train_pred))
print('Précision sur l\'ensemble de test : ', accuracy_score(y_test, y_test_pred))

HTML(pd.DataFrame(confusion_matrix(y_test, y_test_pred),
             index = model_rfc.classes_ + " REAL",
             columns = model_rfc.classes_ + " PREDICTED").to_html())


Précision sur l'ensemble d'entraînement :  0.9993674889310563
Précision sur l'ensemble de test :  0.7415682967959528


In [95]:
print(classification_report(y_test, y_test_pred))

                    precision    recall  f1-score   support

Blessé hospitalisé       0.61      0.47      0.53       562
      Blessé léger       0.77      0.92      0.84      1559
           Indemne       0.70      0.32      0.44       167
               Tué       0.56      0.06      0.11        84

          accuracy                           0.74      2372
         macro avg       0.66      0.44      0.48      2372
      weighted avg       0.72      0.74      0.71      2372



### Recherche des meilleurs modèles

In [96]:
lazy = LazyClassifier(verbose=0, ignore_warnings=True, custom_metric=None)
models, predictions = lazy.fit(X_train, X_test, y_train, y_test)

 97%|█████████▋| 30/31 [00:49<00:02,  2.37s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.005324 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 985
[LightGBM] [Info] Number of data points in the train set: 9486, number of used features: 291
[LightGBM] [Info] Start training from score -1.439776
[LightGBM] [Info] Start training from score -0.419638
[LightGBM] [Info] Start training from score -2.656283
[LightGBM] [Info] Start training from score -3.337489


100%|██████████| 31/31 [00:49<00:00,  1.60s/it]


In [32]:
display(models)

Accuracy  Balanced Accuracy ROC AUC  F1 Score  \
Model                                                                          
NearestCentroid                    0.60               0.58    None      0.63   
BernoulliNB                        0.67               0.55    None      0.68   
BaggingClassifier                  0.73               0.50    None      0.72   
LinearDiscriminantAnalysis         0.71               0.49    None      0.70   
LGBMClassifier                     0.73               0.48    None      0.71   
DecisionTreeClassifier             0.67               0.48    None      0.66   
LogisticRegression                 0.72               0.47    None      0.70   
ExtraTreesClassifier               0.75               0.47    None      0.72   
RandomForestClassifier             0.74               0.45    None      0.72   
LinearSVC                          0.72               0.45    None      0.70   
ExtraTreeClassifier                0.66               0.44    None      0.65   
SGDClassifier                      0.67               0.44    None      0.66   
RidgeClassifierCV                  0.72               0.44    None      0.70   
RidgeClassifier                    0.72               0.44    None      0.70   
AdaBoostClassifier                 0.69               0.43    None      0.67   
SVC                                0.72               0.43    None      0.69   
Perceptron                         0.66               0.43    None      0.65   
CalibratedClassifierCV             0.72               0.42    None      0.68   
PassiveAggressiveClassifier        0.65               0.40    None      0.64   
KNeighborsClassifier               0.69               0.37    None      0.65   
LabelSpreading                     0.37               0.34    None      0.33   
LabelPropagation                   0.37               0.34    None      0.33   
GaussianNB                         0.09               0.33    None      0.03   
QuadraticDiscriminantAnalysis      0.16               0.32    None      0.17   
DummyClassifier                    0.66               0.25    None      0.52   

                               Time Taken  
Model                                      
NearestCentroid                      0.13  
BernoulliNB                          0.15  
BaggingClassifier                    2.08  
LinearDiscriminantAnalysis           0.40  
LGBMClassifier                       0.53  
DecisionTreeClassifier               0.41  
LogisticRegression                   0.28  
ExtraTreesClassifier                 2.26  
RandomForestClassifier               1.87  
LinearSVC                            2.47  
ExtraTreeClassifier                  0.14  
SGDClassifier                        3.57  
RidgeClassifierCV                    0.31  
RidgeClassifier                      0.17  
AdaBoostClassifier                   1.58  
SVC                                 15.34  
Perceptron                           0.23  
CalibratedClassifierCV              12.73  
PassiveAggressiveClassifier          0.35  
KNeighborsClassifier                 0.43  
LabelSpreading                       3.52  
LabelPropagation                     2.81  
GaussianNB                           0.17  
QuadraticDiscriminantAnalysis        0.31  
DummyClassifier                      0.11

## Target en 2 catégories (grave ou léger)

In [97]:
y2 = df_cyclistes_seuls['gravity'].apply(lambda x : 'grave' if x in ("Blessé hospitalisé", "Tué") else 'léger')
y2

0        grave
2        léger
4        grave
5        léger
7        grave
         ...  
21538    grave
21539    léger
21542    léger
21543    grave
21546    léger
Name: gravity, Length: 11858, dtype: object

In [98]:
X_train2, X_test2, y_train2, y_test2 = train_test_split(X_scaled, y2, random_state=36, train_size = 0.8, stratify=y2)

print('Proportion des gravités dans l\'ensemble d\'entraînement : ')
display(y_train2.value_counts(normalize=True))
print('Proportion des gravités dans l\'ensemble de test : ')
display(y_test2.value_counts(normalize=True))

Proportion des gravités dans l'ensemble d'entraînement : 


gravity
léger   0.73
grave   0.27
Name: proportion, dtype: float64

Proportion des gravités dans l'ensemble de test : 


gravity
léger   0.73
grave   0.27
Name: proportion, dtype: float64

### Modèle LogisticRegression

In [99]:
model_rl = LogisticRegression(random_state=36)

# Entraîner le modèle sur les données d'entraînement.
model_rl.fit(X_train2, y_train2)

# Faire des prédictions sur les ensembles d'entraînement et de test.
y_train_pred2 = model_rl.predict(X_train2)
y_test_pred2 = model_rl.predict(X_test2)

# Calculer et afficher la précision (accuracy) sur les deux ensembles.
print('Précision sur l\'ensemble d\'entraînement : ', accuracy_score(y_train2, y_train_pred2))
print('Précision sur l\'ensemble de test : ', accuracy_score(y_test2, y_test_pred2))

HTML(pd.DataFrame(confusion_matrix(y_test2, y_test_pred2),
             index = model_rl.classes_ + " REAL",
             columns = model_rl.classes_ + " PREDICTED").to_html())

Précision sur l'ensemble d'entraînement :  0.8133038161501159
Précision sur l'ensemble de test :  0.7896290050590219


In [100]:
print(classification_report(y_test2, y_test_pred2))

              precision    recall  f1-score   support

       grave       0.65      0.50      0.56       646
       léger       0.83      0.90      0.86      1726

    accuracy                           0.79      2372
   macro avg       0.74      0.70      0.71      2372
weighted avg       0.78      0.79      0.78      2372



In [101]:
model_rfc = RandomForestClassifier(random_state=36)

# Entraîner le modèle sur les données d'entraînement.
model_rfc.fit(X_train2, y_train2)

# Faire des prédictions sur les ensembles d'entraînement et de test.
y_train_pred2 = model_rfc.predict(X_train2)
y_test_pred2 = model_rfc.predict(X_test2)

# Calculer et afficher la précision (accuracy) sur les deux ensembles.
print('Précision sur l\'ensemble d\'entraînement : ', accuracy_score(y_train2, y_train_pred2))
print('Précision sur l\'ensemble de test : ', accuracy_score(y_test2, y_test_pred2))

HTML(pd.DataFrame(confusion_matrix(y_test2, y_test_pred2),
             index = model_rl.classes_ + " REAL",
             columns = model_rl.classes_ + " PREDICTED").to_html())

Précision sur l'ensemble d'entraînement :  0.9994729074425469
Précision sur l'ensemble de test :  0.8115514333895447


In [102]:
print(classification_report(y_test2, y_test_pred2))

              precision    recall  f1-score   support

       grave       0.72      0.50      0.59       646
       léger       0.83      0.93      0.88      1726

    accuracy                           0.81      2372
   macro avg       0.78      0.71      0.73      2372
weighted avg       0.80      0.81      0.80      2372



### Recherche des meilleurs modèles

In [103]:
lazy = LazyClassifier(verbose=0, ignore_warnings=True, custom_metric=None)
models, predictions = lazy.fit(X_train2, X_test2, y_train2, y_test2)

100%|██████████| 31/31 [00:40<00:00,  1.32s/it]

[LightGBM] [Info] Number of positive: 6901, number of negative: 2585
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.003310 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 997
[LightGBM] [Info] Number of data points in the train set: 9486, number of used features: 295
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.727493 -> initscore=0.981941
[LightGBM] [Info] Start training from score 0.981941


In [104]:
display(models)

Accuracy  Balanced Accuracy ROC AUC  F1 Score  \
Model                                                                          
NearestCentroid                    0.77               0.73    None      0.77   
BernoulliNB                        0.77               0.73    None      0.77   
BaggingClassifier                  0.79               0.73    None      0.79   
RandomForestClassifier             0.81               0.72    None      0.80   
ExtraTreesClassifier               0.81               0.72    None      0.80   
LGBMClassifier                     0.80               0.71    None      0.79   
LinearDiscriminantAnalysis         0.79               0.70    None      0.78   
LinearSVC                          0.79               0.70    None      0.78   
LogisticRegression                 0.79               0.70    None      0.78   
SVC                                0.80               0.69    None      0.78   
RidgeClassifierCV                  0.79               0.69    None      0.78   
RidgeClassifier                    0.79               0.69    None      0.78   
GaussianNB                         0.74               0.68    None      0.74   
AdaBoostClassifier                 0.79               0.68    None      0.77   
CalibratedClassifierCV             0.79               0.68    None      0.78   
DecisionTreeClassifier             0.76               0.68    None      0.75   
ExtraTreeClassifier                0.75               0.67    None      0.75   
NuSVC                              0.78               0.66    None      0.76   
SGDClassifier                      0.74               0.66    None      0.74   
KNeighborsClassifier               0.76               0.65    None      0.74   
Perceptron                         0.72               0.65    None      0.72   
PassiveAggressiveClassifier        0.73               0.64    None      0.73   
LabelSpreading                     0.41               0.58    None      0.36   
LabelPropagation                   0.41               0.58    None      0.36   
QuadraticDiscriminantAnalysis      0.67               0.53    None      0.64   
DummyClassifier                    0.73               0.50    None      0.61   

                               Time Taken  
Model                                      
NearestCentroid                      0.17  
BernoulliNB                          0.16  
BaggingClassifier                    2.07  
RandomForestClassifier               1.63  
ExtraTreesClassifier                 2.15  
LGBMClassifier                       0.30  
LinearDiscriminantAnalysis           0.31  
LinearSVC                            0.44  
LogisticRegression                   0.19  
SVC                                 10.37  
RidgeClassifierCV                    0.30  
RidgeClassifier                      0.15  
GaussianNB                           0.15  
AdaBoostClassifier                   1.69  
CalibratedClassifierCV               1.42  
DecisionTreeClassifier               0.37  
ExtraTreeClassifier                  0.14  
NuSVC                               11.21  
SGDClassifier                        0.50  
KNeighborsClassifier                 0.28  
Perceptron                           0.15  
PassiveAggressiveClassifier          0.17  
LabelSpreading                       3.35  
LabelPropagation                     2.54  
QuadraticDiscriminantAnalysis        0.27  
DummyClassifier                      0.11

## Features simples

In [105]:
features_cat = ['obs', 'obsm', 'choc', 'manv', 'catu', 'secu1', 'lum', 'agg', 'int', 'atm',
             'col', 'catr', 'circ', 'prof', 'plan', 'surf', 'infra', 'situ', ]
features_total = features_cat + ['nbv', 'vma']

df_encoded = pd.get_dummies(df_cyclistes_seuls[features_total],columns=features_cat)
X = df_encoded
y = df_cyclistes_seuls['gravity']

standard_scaler = StandardScaler()
X_scaled = pd.DataFrame(standard_scaler.fit_transform(X), columns=X.columns)
X_scaled.head()

nbv   vma  obs_0  obs_1  obs_2  obs_3  obs_4  obs_5  obs_6  obs_7  ...  \
0 -0.14  2.04   0.29  -0.18  -0.04  -0.03  -0.03  -0.03  -0.05  -0.03  ...   
1 -0.14  0.15   0.29  -0.18  -0.04  -0.03  -0.03  -0.03  -0.05  -0.03  ...   
2 -1.94 -1.12   0.29  -0.18  -0.04  -0.03  -0.03  -0.03  -0.05  -0.03  ...   
3 -1.94 -1.12   0.29  -0.18  -0.04  -0.03  -0.03  -0.03  -0.05  -0.03  ...   
4 -0.14 -1.12  -3.46  -0.18  -0.04  -0.03  -0.03  -0.03  -0.05  -0.03  ...   

   infra_6  infra_8  infra_9  situ_1  situ_2  situ_3  situ_4  situ_5  situ_6  \
0    -0.11    -0.08    -0.18    0.51   -0.03   -0.11   -0.14   -0.41   -0.14   
1    -0.11    -0.08    -0.18    0.51   -0.03   -0.11   -0.14   -0.41   -0.14   
2    -0.11    -0.08    -0.18    0.51   -0.03   -0.11   -0.14   -0.41   -0.14   
3    -0.11    -0.08    -0.18    0.51   -0.03   -0.11   -0.14   -0.41   -0.14   
4    -0.11    -0.08    -0.18    0.51   -0.03   -0.11   -0.14   -0.41   -0.14   

   situ_8  
0   -0.11  
1   -0.11  
2   -0.11  
3   -0.11  
4   -0.11  

[5 rows x 150 columns]

In [106]:
y2 = df_cyclistes_seuls['gravity'].apply(lambda x : 'grave' if x in ("Blessé hospitalisé", "Tué") else 'léger')

X_train2, X_test2, y_train2, y_test2 = train_test_split(X_scaled, y2, random_state=36, train_size = 0.8, stratify=y2)

print('Proportion des gravités dans l\'ensemble d\'entraînement : ')
display(y_train2.value_counts(normalize=True))
print('Proportion des gravités dans l\'ensemble de test : ')
display(y_test2.value_counts(normalize=True))

Proportion des gravités dans l'ensemble d'entraînement : 


gravity
léger   0.73
grave   0.27
Name: proportion, dtype: float64

Proportion des gravités dans l'ensemble de test : 


gravity
léger   0.73
grave   0.27
Name: proportion, dtype: float64

### Modèle LogisticRegression

In [107]:
model_rl = LogisticRegression(random_state=36)

# Entraîner le modèle sur les données d'entraînement.
model_rl.fit(X_train2, y_train2)

# Faire des prédictions sur les ensembles d'entraînement et de test.
y_train_pred2 = model_rl.predict(X_train2)
y_test_pred2 = model_rl.predict(X_test2)

# Calculer et afficher la précision (accuracy) sur les deux ensembles.
print('Précision sur l\'ensemble d\'entraînement : ', accuracy_score(y_train2, y_train_pred2))
print('Précision sur l\'ensemble de test : ', accuracy_score(y_test2, y_test_pred2))

HTML(pd.DataFrame(confusion_matrix(y_test2, y_test_pred2),
             index = model_rl.classes_ + " REAL",
             columns = model_rl.classes_ + " PREDICTED").to_html())

Précision sur l'ensemble d'entraînement :  0.7808349146110057
Précision sur l'ensemble de test :  0.7786677908937606


In [108]:
print(classification_report(y_test2, y_test_pred2))

              precision    recall  f1-score   support

       grave       0.65      0.40      0.50       646
       léger       0.80      0.92      0.86      1726

    accuracy                           0.78      2372
   macro avg       0.73      0.66      0.68      2372
weighted avg       0.76      0.78      0.76      2372



### Modèle RandomForestClassifier

In [109]:
model_rfc = RandomForestClassifier(random_state=36)

# Entraîner le modèle sur les données d'entraînement.
model_rfc.fit(X_train2, y_train2)

# Faire des prédictions sur les ensembles d'entraînement et de test.
y_train_pred2 = model_rfc.predict(X_train2)
y_test_pred2 = model_rfc.predict(X_test2)

# Calculer et afficher la précision (accuracy) sur les deux ensembles.
print('Précision sur l\'ensemble d\'entraînement : ', accuracy_score(y_train2, y_train_pred2))
print('Précision sur l\'ensemble de test : ', accuracy_score(y_test2, y_test_pred2))

HTML(pd.DataFrame(confusion_matrix(y_test2, y_test_pred2),
             index = model_rl.classes_ + " REAL",
             columns = model_rl.classes_ + " PREDICTED").to_html())

Précision sur l'ensemble d'entraînement :  0.9808138309087076
Précision sur l'ensemble de test :  0.7837268128161888


In [110]:
print(classification_report(y_test2, y_test_pred2))

              precision    recall  f1-score   support

       grave       0.65      0.46      0.53       646
       léger       0.82      0.91      0.86      1726

    accuracy                           0.78      2372
   macro avg       0.73      0.68      0.70      2372
weighted avg       0.77      0.78      0.77      2372



### Prédiction sur exemple précis :

In [111]:
feature_cat = ['obs', 'obsm', 'choc', 'manv',  'catu', 'secu1', 'lum', 'agg', 'int', 'atm',
             'col', 'catr', 'circ', 'prof', 'plan', 'surf', 'infra', 'situ', ]
features_total = features_cat + ['nbv', 'vma']

data_cycl = {'obs':[0], 'obsm':[0], 'choc': [0], 'manv':[1], 'catu':[1], 'secu1':[2], 'lum':[1],
             'agg':[1], 'int': [1], 'atm':[1], 'col':[7], 'catr':[3], 'circ':[2], 'prof':[1], 'plan':[1],
             'surf':[1], 'infra':[0], 'situ':[1], 'nbv':[1], 'vma':[80]}
X_cycl = pd.DataFrame(data=data_cycl, index=[0])
X_cycl_encoded= pd.get_dummies(X_cycl, columns=features_cat)

# Get missing columns in the training test
missing_cols = set( df_encoded.columns ) - set( X_cycl_encoded.columns )
# Add a missing column in test set with default value equal to 0
for c in missing_cols:
    X_cycl_encoded[c] = False
# Ensure the order of column in the test set is in the same order than in train set
X_cycl_encoded = X_cycl_encoded[df_encoded.columns]


X_cycl_scaled = pd.DataFrame(standard_scaler.transform(X_cycl_encoded), columns=X_cycl_encoded.columns)

y_cycl_predict = model_rfc.predict(X_cycl_scaled)
y_cycl_predict

array(['grave'], dtype=object)

### Recherche des meilleurs modèles

In [55]:
lazy = LazyClassifier(verbose=0, ignore_warnings=True, custom_metric=None)
models, predictions = lazy.fit(X_train2, X_test2, y_train2, y_test2)

100%|██████████| 31/31 [00:25<00:00,  1.23it/s]

[LightGBM] [Info] Number of positive: 6901, number of negative: 2585
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001446 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 385
[LightGBM] [Info] Number of data points in the train set: 9486, number of used features: 122
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.727493 -> initscore=0.981941
[LightGBM] [Info] Start training from score 0.981941


In [56]:
display(models)

Accuracy  Balanced Accuracy ROC AUC  F1 Score  \
Model                                                                          
NearestCentroid                    0.75               0.70    None      0.75   
BernoulliNB                        0.75               0.70    None      0.75   
BaggingClassifier                  0.76               0.68    None      0.76   
RandomForestClassifier             0.78               0.68    None      0.77   
ExtraTreesClassifier               0.77               0.68    None      0.76   
LGBMClassifier                     0.78               0.67    None      0.77   
LinearDiscriminantAnalysis         0.78               0.66    None      0.76   
LogisticRegression                 0.78               0.66    None      0.76   
LinearSVC                          0.78               0.66    None      0.76   
RidgeClassifierCV                  0.78               0.66    None      0.76   
RidgeClassifier                    0.78               0.66    None      0.76   
CalibratedClassifierCV             0.78               0.66    None      0.76   
AdaBoostClassifier                 0.77               0.65    None      0.75   
SGDClassifier                      0.77               0.65    None      0.75   
KNeighborsClassifier               0.75               0.65    None      0.74   
ExtraTreeClassifier                0.71               0.64    None      0.71   
SVC                                0.78               0.64    None      0.75   
LabelPropagation                   0.63               0.64    None      0.65   
LabelSpreading                     0.63               0.64    None      0.65   
DecisionTreeClassifier             0.70               0.63    None      0.71   
NuSVC                              0.77               0.62    None      0.74   
Perceptron                         0.71               0.62    None      0.71   
PassiveAggressiveClassifier        0.68               0.60    None      0.68   
QuadraticDiscriminantAnalysis      0.38               0.55    None      0.33   
GaussianNB                         0.35               0.54    None      0.28   
DummyClassifier                    0.73               0.50    None      0.61   

                               Time Taken  
Model                                      
NearestCentroid                      0.07  
BernoulliNB                          0.09  
BaggingClassifier                    0.78  
RandomForestClassifier               1.27  
ExtraTreesClassifier                 1.33  
LGBMClassifier                       0.18  
LinearDiscriminantAnalysis           0.14  
LogisticRegression                   0.09  
LinearSVC                            0.48  
RidgeClassifierCV                    0.19  
RidgeClassifier                      0.13  
CalibratedClassifierCV               1.30  
AdaBoostClassifier                   0.94  
SGDClassifier                        0.38  
KNeighborsClassifier                 0.18  
ExtraTreeClassifier                  0.08  
SVC                                  5.48  
LabelPropagation                     2.46  
LabelSpreading                       3.28  
DecisionTreeClassifier               0.18  
NuSVC                                5.54  
Perceptron                           0.15  
PassiveAggressiveClassifier          0.10  
QuadraticDiscriminantAnalysis        0.17  
GaussianNB                           0.08  
DummyClassifier                      0.07

# Dataset total

## Target en 4 catégories

In [41]:
df_encoded_total = pd.get_dummies(df_total[features_total],columns=features_cat)
X_total = df_encoded_total
y_total = df_total['gravity']

standard_scaler = StandardScaler()
X_scaled_total = pd.DataFrame(standard_scaler.fit_transform(X_total), columns=X_total.columns)

X_train, X_test, y_train, y_test = train_test_split(X_scaled_total, y_total, random_state=36, train_size = 0.8, stratify=y_total)

print('Proportion des gravités dans l\'ensemble d\'entraînement : ')
display(y_train.value_counts(normalize=True))
print('Proportion des gravités dans l\'ensemble de test : ')
display(y_test.value_counts(normalize=True))

Proportion des gravités dans l'ensemble d'entraînement : 


gravity
Indemne              0.42
Blessé léger         0.40
Blessé hospitalisé   0.15
Tué                  0.03
Name: proportion, dtype: float64

Proportion des gravités dans l'ensemble de test : 


gravity
Indemne              0.42
Blessé léger         0.40
Blessé hospitalisé   0.15
Tué                  0.03
Name: proportion, dtype: float64

### Modèle LogisticRegression

In [42]:
model_rl = LogisticRegression(random_state=36)

# Entraîner le modèle sur les données d'entraînement.
model_rl.fit(X_train, y_train)

# Faire des prédictions sur les ensembles d'entraînement et de test.
y_train_pred = model_rl.predict(X_train)
y_test_pred = model_rl.predict(X_test)

# Calculer et afficher la précision (accuracy) sur les deux ensembles.
print('Précision sur l\'ensemble d\'entraînement : ', accuracy_score(y_train, y_train_pred))
print('Précision sur l\'ensemble de test : ', accuracy_score(y_test, y_test_pred))

HTML(pd.DataFrame(confusion_matrix(y_test, y_test_pred),
             index = model_rl.classes_ + " REAL",
             columns = model_rl.classes_ + " PREDICTED").to_html())

Précision sur l'ensemble d'entraînement :  0.6635244087029047
Précision sur l'ensemble de test :  0.6626989847402129


In [46]:
coef_df = pd.DataFrame({'Features' : X_total.columns, 'Coefficient': model_rl.coef_[0]})
coef_df = coef_df.sort_values('Coefficient', ascending=False)
HTML(coef_df.to_html())

### Modèle RandomForestClassifier

In [165]:
model_rfc = RandomForestClassifier(random_state=36)

# Entraîner le modèle sur les données d'entraînement.
model_rfc.fit(X_train, y_train)

# Faire des prédictions sur les ensembles d'entraînement et de test.
y_train_pred = model_rfc.predict(X_train)
y_test_pred = model_rfc.predict(X_test)

# Calculer et afficher la précision (accuracy) sur les deux ensembles.
print('Précision sur l\'ensemble d\'entraînement : ', accuracy_score(y_train, y_train_pred))
print('Précision sur l\'ensemble de test : ', accuracy_score(y_test, y_test_pred))

HTML(pd.DataFrame(confusion_matrix(y_test, y_test_pred),
             index = model_rfc.classes_ + " REAL",
             columns = model_rfc.classes_ + " PREDICTED").to_html())

Précision sur l'ensemble d'entraînement :  0.9996499068134312
Précision sur l'ensemble de test :  0.7094668341605058


### GridSearch

In [174]:
rfc=RandomForestClassifier(random_state=36)

param_grid = { 
    'n_estimators': [200, 500],
    'max_features': ['auto', 'sqrt', 'log2'],
    'max_depth' : [4,5,6,7,8],
    'criterion' :['gini', 'entropy']
}

gridsearch_rfc = GridSearchCV(estimator=rfc, param_grid=param_grid, cv= 5, scoring="accuracy")

In [175]:
gridsearch_rfc.fit(X_train, y_train)

print('Best params', gridsearch_rfc.best_params_)
print('Best score', gridsearch_rfc.best_score_)

best_model = gridsearch_rfc.best_estimator_
best_model.fit(X_train, y_train)

y_train_pred_best = best_model.predict(X_train)
y_test_pred_best = best_model.predict(X_test)

print(accuracy_score(y_train, y_train_pred_best))
print(accuracy_score(y_test, y_test_pred_best))

Best params {'criterion': 'entropy', 'max_depth': 8, 'max_features': 'sqrt', 'n_estimators': 200}
Best score 0.6265329356141216
0.6285305353336696
0.6259395786568916


### Recherche des meilleurs modèles

In [99]:
lazy = LazyClassifier(verbose=0, ignore_warnings=True, custom_metric=None)
models, predictions = lazy.fit(X_train, X_test, y_train, y_test)

 87%|████████▋ | 27/31 [00:45<00:18,  4.58s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.003354 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 994
[LightGBM] [Info] Number of data points in the train set: 9486, number of used features: 294
[LightGBM] [Info] Start training from score -1.439776
[LightGBM] [Info] Start training from score -0.419638
[LightGBM] [Info] Start training from score -2.656283
[LightGBM] [Info] Start training from score -3.337489


100%|██████████| 31/31 [00:45<00:00,  1.48s/it]


# PREDICTIONS

 - garder moins de features ? (afficher importance des features)
 - choisir un modèle et faire des prédictions pour une personne avec certaines caractéristiques précises (possibilité de choisir

In [70]:
options = {'obs': {1 : 'Véhicule en stationnement', 2 : 'Arbre', 3 : 'Glissière métallique', 4 : 'Glissière béton', 5 : 'Autre glissière', 6 : 'Bâtiment, mur, pile de pont',
                             7 : 'Support de signalisation verticale ou poste d’appel d’urgence', 8 : 'Poteau', 9 : 'Mobilier urbain', 10 : 'Parapet', 11 : 'Ilot, refuge, borne haute',
                             12 : 'Bordure de trottoir', 13 : 'Fossé, talus, paroi rocheuse', 14 : 'Autre obstacle fixe sur chaussée', 15 : 'Autre obstacle fixe sur trottoir ou accotement',
                             16 : 'Sortie de chaussée sans obstacle', 17 : 'Buse – tête d’aqueduc'},
                'obsm':{0 : 'Aucun', 1 : 'Piéton', 2 : 'Véhicule', 4 : 'Véhicule sur rail', 5 : 'Animal domestique', 6 : 'Animal sauvage', 9 : 'Autre'},
                'choc': {0 : 'Aucun', 1 : 'Avant', 2 : 'Avant droit', 3 : 'Avant gauche', 4 : 'Arrière', 5 : 'Arrière droit', 6 : 'Arrière gauche', 7 : 'Côté droit', 8 : 'Côté gauche', 9 : 'Chocs multiples (tonneaux)'},
                'manv':{1 : 'Sans changement de direction', 2 : 'Même sens, même file', 3 : 'Entre 2 files', 4 : 'En marche arrière', 5 : 'A contresens', 6 : 'En franchissant le terre-plein central',
                              7 : 'Dans le couloir bus, dans le même sens', 8 : 'Dans le couloir bus, dans le sens inverse', 9 : 'En s’insérant', 10 : 'En faisant demi-tour sur la chaussée',
                              11 : 'Changeant de file à gauche', 12 : 'Changeant de file à droite', 13 : 'Déporté à gauche', 14 : 'Déporté à droite',  15 : 'Tournant à gauche', 16 :  'Tournant à droite',
                              17 : 'Dépassant à gauche', 18 : 'Dépassant à droite', 19 : 'Traversant la chaussée', 20 : 'Manœuvre de stationnement', 21 : 'Manœuvre d’évitement',
                              22 : 'Ouverture de porte', 23 : 'Arrêté (hors stationnement)', 24 : 'En stationnement (avec occupants)', 25 : 'Circulant sur trottoir', 26 : 'Autres manœuvres'},
                'catu': {1 : 'Conducteur', 2 : 'Passager', 3 : 'Piéton'},
                'secu1' : {0 : 'Aucun équipement', 1 : 'Ceinture', 2 : 'Casque', 3 : 'Dispositif enfants', 4 : 'Gilet réfléchissant', 5 : 'Airbag (2RM/3RM)', 6  : 'Gants (2RM/3RM)',
                             7 : 'Gants + Airbag (2RM/3RM)', 8 : 'Non déterminable',  9 : 'Autre'},
                'lum': {1 :'Plein jour', 2 :'Crépuscule ou aube', 3 : 'Nuit sans éclairage public', 4 : 'Nuit avec éclairage public non allumé', 5 : 'Nuit avec éclairage public allumé'},
                'agg': {1 : 'Hors agglomération', 2 : 'En agglomération'},
                'int': {1 : 'Hors intersection', 2 : 'Intersection en X', 3 : 'Intersection en T', 4 : 'Intersection en Y', 5 : 'Intersection à plus de 4 branches', 6 : 'Giratoire', 7 : 'Place', 8 : 'Passage à niveau', 9 : 'Autre intersection'},
                'atm' : {1  : 'Normale', 2 : 'Pluie légère', 3 : 'Pluie forte', 4 : 'Neige - grêle', 5 : 'Brouillard - fumée', 6 : 'Vent fort - tempête', 7 : 'Temps éblouissant', 8 : 'Temps couvert', 9 : 'Autre'},
                'col': {1 : 'Deux véhicules - frontale', 2 : 'Deux véhicules – par l’arrière', 3 : 'Deux véhicules – par le coté', 4 : 'Trois véhicules et plus – en chaîne', 5 : 'Trois véhicules et plus - collisions multiples', 6 : 'Autre collision', 7 : 'Sans collision' },
                'catr': {1:'Autoroute', 2: 'Route nationale', 3:'Route Départementale', 4:'Voie Communale', 5:'Hors réseau public', 6:'Parc de stationnement ouvert à la circulation publique', 7:'Routes de métropole urbaine', 9:'autre'},
                'circ':{1:'Sens unique', 2:'Bidirectionnelle', 3:'À chaussées séparées', 4:'Voies d\'affectation variable'},
                'prof': {1:'Plat', 2:'Pente', 3 :'Sommet de côte', 4:'Bas de côte'},
                'plan': {1 :'Partie rectiligne', 2:'En courbe à gauche', 3:'En courbe à droite', 4:'En « S »'},
                'surf': {1:'Normale', 2:'Mouillée', 3:'Flaques', 4 :'Inondée', 5:'Enneigée', 6:'Boue', 7:'Verglacée', 8:'Corps gras – huile', 9 :'Autre'},
                'infra':{0 :'Aucun', 1:'Souterrain - tunnel' , 2:'Pont - autopont', 3: 'Bretelle d’échangeur ou de raccordement', 4: 'Voie ferrée', 5:'Carrefour aménagé', 6 :'Zone piétonne', 7 :'Zone de péage', 8:'Chantier', 9 :'Autres'},
                 'situ':{1:'Sur chaussée',  2:'Sur bande d’arrêt d’urgence', 3:'Sur accotement', 4:'Sur trottoir', 5:'Sur piste cyclable', 6:'Sur autre voie spéciale'}
                 }

In [75]:
for k in options.keys():
    liste = [str(k)+' - ' + str(v) for k,v in options[k].items()]
    print(liste)

['1 -- Véhicule en stationnement', '2 -- Arbre', '3 -- Glissière métallique', '4 -- Glissière béton', '5 -- Autre glissière', '6 -- Bâtiment, mur, pile de pont', '7 -- Support de signalisation verticale ou poste d’appel d’urgence', '8 -- Poteau', '9 -- Mobilier urbain', '10 -- Parapet', '11 -- Ilot, refuge, borne haute', '12 -- Bordure de trottoir', '13 -- Fossé, talus, paroi rocheuse', '14 -- Autre obstacle fixe sur chaussée', '15 -- Autre obstacle fixe sur trottoir ou accotement', '16 -- Sortie de chaussée sans obstacle', '17 -- Buse – tête d’aqueduc']
['0 -- Aucun', '1 -- Piéton', '2 -- Véhicule', '4 -- Véhicule sur rail', '5 -- Animal domestique', '6 -- Animal sauvage', '9 -- Autre']
['0 -- Aucun', '1 -- Avant', '2 -- Avant droit', '3 -- Avant gauche', '4 -- Arrière', '5 -- Arrière droit', '6 -- Arrière gauche', '7 -- Côté droit', '8 -- Côté gauche', '9 -- Chocs multiples (tonneaux)']
['1 -- Sans changement de direction', '2 -- Même sens, même file', '3 -- Entre 2 files', '4 -- En 